sous-chef/sous_chef/pantry_list/read_pantry_list.py

```diff
@@ -25,68 +25,139 @@ def __init__(self, config: DictConfig, gsheets_helper: GsheetsHelper):
+
     @staticmethod
-    def _get_pluralized_form(row: Series):
-        if row.plural_ending in ["ies", "ves"]:
-            return row.ingredient[:-1] + row.plural_ending
+    def _get_pluralized_form(plural_ending: str, ingredient: str):
+        if plural_ending in ["ies", "ves"]:
+            return ingredient[:-1] + plural_ending
         else:
-            return row.ingredient + row.plural_ending
+            return ingredient + plural_ending
 
     def _load_complex_pantry_list_for_search(self):
-        basic_pantry_list = self._get_basic_pantry_list()
-        misspelled_pantry_list = self._retrieve_misspelled_pantry_list()
-        plural_pantry_list = self._retrieve_plural_pantry_list()
         return pd.concat(
-            [basic_pantry_list, misspelled_pantry_list, plural_pantry_list]
+            [
+                self._get_basic_pantry_list(),
+                self._retrieve_misspelled_pantry_list(),
+                self._get_replacement_pantry_list(),
+            ]
         )
 
     def _retrieve_basic_pantry_list(self) -> DataFrame:
         dataframe = self.gsheets_helper.get_worksheet(
             self.config.workbook_name, self.config.ingredient_sheet_name
         )
         dataframe["item_plural"] = dataframe.apply(
-            self._get_pluralized_form, axis=1
+            lambda x: self._get_pluralized_form(x.plural_ending, x.ingredient),
+            axis=1,
         )
         return dataframe
 
     def _retrieve_misspelled_pantry_list(self) -> DataFrame:
         misspelled_list = self.gsheets_helper.get_worksheet(
             self.config.workbook_name, self.config.misspelling_sheet_name
         )
-        misspelled_list["label"] = "misspelled_form"
 
+        mask_and_replaced = misspelled_list.replacement_ingredient != ""
+        # set up misspelled ingredients that are not replaced
+        without_replacement = misspelled_list[~mask_and_replaced][
+            ["misspelled_ingredient", "true_ingredient"]
+        ]
+        without_replacement["label"] = "misspelled_form"
+        without_replacement["replace_factor"] = 1
+        without_replacement["replace_unit"] = ""
+        # set up misspelled ingredients that are replaced
+        with_replacement = misspelled_list[mask_and_replaced]
+        with_replacement["label"] = "misspelled_replaced_form"
+        with_replacement = pd.merge(
+            self.replacement_pantry_list,
+            with_replacement,
+            how="inner",
+            on=["replacement_ingredient", "true_ingredient"],
+        )
+        with_replacement = with_replacement[
+            [
+                "misspelled_ingredient",
+                "true_ingredient",
+                "label",
+                "replace_factor",
+                "replace_unit",
+            ]
+        ]
+
+        # join all misspelled ingredients with basic pantry list
+        misspelled_list = pd.concat([without_replacement, with_replacement])
```

Comment:
When we have a merge with inner, do we want to toss an error afterwards if the shape is less than before the merge?

Reply:
Good idea. This has been added.